[Feature] Add support for `endBlock` in data sources

incrypto32 opened the pull request.

incrypto32 commented on core/src/subgraph/runner.rs:
```diff
@@ -144,12 +144,59 @@ where
+                        // Filter out data sources that have reached their end block.
+                        .filter(|ds| match ds.end_block() {
+                            Some(end_block) => {
+                                current_ptr.as_ref().map(|ptr| ptr.number).unwrap_or(0) > end_block
```
@leoyvens we need to somehow know that the block is final before the end_block reached datasource is removed when reconstructing the filter, how do we know which block is final and which is not at this point?

mangas commented on chain/ethereum/src/data_source.rs:
```diff
@@ -217,6 +225,8 @@ impl blockchain::DataSource<Chain> for DataSource {
             address,
             start_block: 0,
+            // EBTODO: Re-evaluate if this needs to be set to done_at,
+            end_block: done_at,
```
As far as I understand, done_at is only used for file data sources, it does not make sense to use it here as there is no "done" for a Dynamic Data Source in other scenarios. Would be good to confirm with @leoyvens as he may have intended this to be used in some other and add a comment to explain what it is and what it is used for to avoid confusion in the future.

mangas replied:
This is to say I don't think we should use this field as that will make it quite difficult to understand the meaning on different contexts, I'd rather just add an explicit field end_block or maybe consider deleting the data source once it's no longer needed?

incrypto32 replied:
I just set the field to `None` for now, but if we decide to have a way to have an `endBlock` for dynamic data sources, we would need to use `done_at` at some point and then set `endBlock` here.

mangas commented on core/src/subgraph/context.rs:
```diff
@@ -177,10 +174,13 @@ impl<C: Blockchain, T: RuntimeHostBuilder<C>> IndexingContext<C, T> {
         &self.instance
     }
+
+    pub fn hosts(&self) -> &[Arc<T::Host>] {
```
There's already an instance fn, why not just make the field public?

incrypto32 replied:
done

mangas commented on core/src/subgraph/context/instance.rs:
```diff
@@ -176,6 +180,7 @@ where
             return vec![];
         }
 
+        // EBTODO
```
??

mangas commented on chain/ethereum/src/data_source.rs:
```diff
@@ -146,6 +152,8 @@ impl blockchain::DataSource<Chain> for DataSource {
             address,
             mapping,
             context,
+            // EBTODO: Re-evaluate if endBlock need to be considered
```
no need for this comment

mangas commented on chain/ethereum/src/data_source.rs:
```diff
@@ -217,6 +225,8 @@ impl blockchain::DataSource<Chain> for DataSource {
             manifest_idx,
             address,
             start_block: 0,
+            // EBTODO: Re-evaluate if this needs to be set to done_at,
```
comment

mangas commented on core/src/subgraph/context/instance.rs:
```diff
@@ -215,10 +220,17 @@ where
         &self.hosts.hosts()
     }
+
+    pub fn static_data_sources(&self) -> Arc<Vec<DataSource<C>>> {
```
Why?

incrypto32 replied:
@mangas I dont understand, are you suggesting making the field public instead of exposing it in a method?

mangas replied:
I don't think we should be adding getter methods unless they are absolutely needed. I'm also not too happy about introducing "static_data_sources" as a term, historically we have referred to these as "data_sources" and "dynamic_data_sources" if we need this distinction then perhaps it should be done with types

incrypto32 replied:
makes sense, so better making it `pub` than exposing it in a method, got it.

Also i think yes we can rename it to `data_sources` instead, This is true for `manifest.data_sources` where only "static" datasources where stored. 

We can keep that naming, i was trying to make a clear distinction because, Although the `SubgraphManifest.data_sources` was meant to be just the "static" datasources, it used to carry both static and dynamic datasources when its mutated to add dynamic sources in `build_subgraph_runner`.

mangas commented on core/src/subgraph/context/instance.rs:
```diff
@@ -215,10 +220,17 @@ where
+        self.static_data_sources.clone()
+    }
+
+    pub fn templates(&self) -> Arc<Vec<DataSourceTemplate<C>>> {
```
same here

mangas commented on core/src/subgraph/context/instance.rs:
```diff
@@ -36,6 +37,8 @@ where
     pub fn from_manifest(
         logger: &Logger,
         manifest: SubgraphManifest<C>,
+        static_data_sources: Vec<DataSource<C>>,
```
the terminology we use is data sources and dynamic data source, please keep this for consistency

incrypto32 replied:
ty for pointing out, actually that vec is not at all needed there since manifest carries it, so i just removed it. the reasoning behind `static_data_sources` naming was that in the same code `data_sources` is used to indicate a vec containing both static and dynamic ds's

mangas commented on core/src/subgraph/context/instance.rs:
```diff
@@ -59,7 +63,7 @@ where
         // we use the same order here as in the subgraph manifest to make the
         // event processing behavior predictable
-        for ds in manifest.data_sources {
+        for ds in data_sources {
```
If manifest already has all the necessary information why pass these separate?

incrypto32 replied:
What manifest has is the static data sources defined in the manifest, earlier the manifest was mutated to add in created dynamic datasources, this is no longer the case, the `data_sources` vec now has the complete static_data_source + dynamic_data_sources in it, and `manifest.data_sources` only contains the static data sources

mangas replied:
We are still passing the templates around, so I'm not sure about this code, if we don't need the entire manifest anymore then could we just pass the bits that are needed?

incrypto32 replied:
The reason for passing `datasources` separately from the manifest was to avoid mutating the manifest to add in the dynamic data sources (which was also a TODO in a comment)

> We are still passing the templates around, so I'm not sure about this code
Templates are passed on to `SubgraphInstance` which is used in `HostBuilder`

The manifest is still used and is needed for passing down `manifest.data_sources` and `manifest.templates` to the Instance.

mangas commented on core/src/subgraph/runner.rs:
```diff
@@ -144,12 +144,59 @@ where
+            // if not enabled we just stick to the filter based on all the data sources.
+            // This specifically removes dynamic data sources based filters because these can be derived
+            // from templates AND this reduces the cost of egress traffic by making the payloads smaller.
+            let filter = if static_filters {
```
Why was this moved to the runner?

incrypto32 replied:
We now need to rebuild the TriggerFilter every time the Blockstream is restarted, this is because  its not possible to remove a datasource from TriggerFilter when its endBlock is reached, As per my discussion with @leoyvens we decided reconstructing the trigger filter is a more robust way to deal with this

mangas commented on graph/src/data_source/mod.rs:
```diff
@@ -158,9 +158,19 @@ impl<C: Blockchain> DataSource<C> {
+                // If the data source has an `endBlock` field, check that the trigger block is
+                // within the range of blocks that the data source is supposed to handle.
+                // Otherwise, ignore the trigger.
+                if let Some(end_block) = ds.end_block() {
```
Maybe this can be added as another pattern with a guard, feels like that would be a lot easier read. something like: 

```
(Self::Onchain(ds), _ ) if ds.has_expired(block.number())
```

incrypto32 replied:
This is done

mangas commented on store/postgres/src/dynds/mod.rs:
```diff
@@ -47,7 +47,7 @@ pub(crate) fn revert(
 }
 
-pub(crate) fn update_offchain_status(
+pub(crate) fn update_processed_datasources_status(
```
Why? What does "processed" add to this name that it didn't have before?

incrypto32 replied:
Ah nice catch, this was done because initially the plan was to use done_at for on-chain data sources as well, but midway we decided that is not neccessary, will change it back

mangas commented on store/test-store/tests/chain/ethereum/manifest.rs:
```diff
@@ -369,6 +369,44 @@ specVersion: 0.0.2
+    let end_block = data_source.as_onchain().unwrap().end_block;
+
+    assert_eq!(Some(9562481), end_block);
+    assert_eq!("Qmmanifest", manifest.id.as_str());
```
What is this actually adding to the test?

incrypto32 replied:
This was just a test i initially wrote to test the parsing of the endBlock field, if you think its not relevant i'll remove it

mangas replied:
No, I'm all for the test, I think that's great, I meant the assertion of Qmmanifest :)

incrypto32 replied:
Yes, it's not relevant was a residue of a copy paste, removed it

mangas commented on chain/near/src/data_source.rs:
```diff
@@ -75,6 +75,13 @@ impl blockchain::DataSource<Chain> for DataSource {
     }
 
+    fn has_expired(&self, block: BlockNumber) -> bool {
+        match self.source.end_block {
```
if you add end_block() to the trait, then this could be implemented in the trait itself as default and wouldn't need to be copied into all the trait implementations. Might not be a bad trade-off

incrypto32 replied:
@mangas then we'd still have to go around implementing `end_block()` on all Datasources right? its either `end_block()` or `has_expired()`. Lets go with `has_expired` then?

mangas commented on core/src/subgraph/runner.rs:
```diff
@@ -144,12 +144,58 @@ where
+            // if not enabled we just stick to the filter based on all the data sources.
+            // This specifically removes dynamic data sources based filters because these can be derived
+            // from templates AND this reduces the cost of egress traffic by making the payloads smaller.
+            let filter = if static_filters {
```
I think this needs to go into some functions rather than being inlined here

incrypto32 commented on core/src/subgraph/runner.rs:
```diff
@@ -657,11 +717,6 @@ where
-        // Merge filters from data sources into the block stream builder
-        self.ctx
-            .filter
-            .extend(data_sources.iter().filter_map(|ds| ds.as_onchain()));
```
We dont need to Merge the filters anymore since we rebuild the filters each time the block stream is restarted. And block stream is restarted everytime an onchain data source is created

incrypto32 commented on core/src/subgraph/runner.rs:
```diff
@@ -1041,7 +1096,22 @@ where
 
         self.revert_state(subgraph_ptr.number)?;
 
-        Ok(Action::Continue)
+        // If any datasource has reached its end block in the range [revert_to_ptr, subgraph_ptr]
+        // then we need to restart the blockstream to reset the TriggerFilter.
+        let needs_restart: bool = self
+            .inputs
+            .end_blocks
+            .range(revert_to_ptr.number..=subgraph_ptr.number)
+            .next()
+            .is_some();
+
+        let action = if needs_restart {
+            Action::Restart
+        } else {
+            Action::Continue
+        };
+
+        Ok(action)
     }
```
This the part where we determine if the blockstream is to be restarted on a reorg. If there is atleast one datasource which reached its endBlock during the block range for which the reorg is happening then we need to restart

incrypto32 commented:
Rebased with master, Addressed the blockstream restart issue, We now restart the block stream on a reorg when there is an endBlock reached datasource in the reorg range

mangas commented on core/src/subgraph/runner.rs:
```diff
@@ -106,6 +106,57 @@ where
+        // if not enabled we just stick to the filter based on all the data sources.
+        // This specifically removes dynamic data sources based filters because these can be derived
+        // from templates AND this reduces the cost of egress traffic by making the payloads smaller.
+        let filter = if static_filters {
```
you can get rid of this nesting, either `if static_filters { ... return filter}` and remove the else or just return the result of the if expression.

mangas replied:
This doesn't seem to be addressed

mangas commented on tests/tests/runner_tests.rs:
```diff
@@ -309,6 +309,52 @@ async fn derived_loaders() {
 }
 
+#[tokio::test]
+async fn end_block() -> anyhow::Result<()> {
```
is it possible to add a test for the revert to make sure we have coverage of the filter rebuild?

incrypto32 replied:
Just made the changes for making the filter rebuild testable

mangas commented on chain/cosmos/src/data_source.rs:
```diff
@@ -505,6 +509,8 @@ pub struct MappingMessageHandler {
 pub struct Source {
     #[serde(rename = "startBlock", default)]
     pub start_block: BlockNumber,
+    #[serde(rename = "endBlock")]
```
https://serde.rs/container-attrs.html#rename_all  <- at this point might as well add this

mangas commented on chain/near/src/data_source.rs:
```diff
@@ -498,5 +502,7 @@ pub(crate) struct Source {
     pub(crate) account: Option<String>,
     #[serde(rename = "startBlock", default)]
     pub(crate) start_block: BlockNumber,
+    #[serde(rename = "endBlock")]
```
prolly same here https://serde.rs/container-attrs.html#rename_all

incrypto32 replied:
done

mangas commented on core/src/subgraph/runner.rs:
```diff
@@ -144,12 +197,17 @@ where
 
             let block_stream_canceler = CancelGuard::new();
             let block_stream_cancel_handle = block_stream_canceler.handle();
+            // TriggerFilter needs to be rebuilt eveytime the blockstream is restarted
```
Exactly this, if the build returns the trigger filter and replace it then we know what mutations it needs to make and the build_filter function can then take a reference instead of mutable ref

incrypto32 replied:
Thanks, just did that

mangas commented on core/src/subgraph/runner.rs:
```diff
@@ -144,12 +197,17 @@ where
-                    .cancelable(&block_stream_canceler, || Err(CancelableError::Cancel));
+            let mut block_stream = new_block_stream(
+                &self.inputs,
+                self.ctx.filter.as_ref().unwrap(), // Safe to unwrap as we just called `build_filter` in the previous line
```
we would also still have the filter here so maybe we could just pass that in instead of unwrapping, mostly a read improvement

incrypto32 replied:
To prevent the unwrapping, i would need to clone the filter to once to assign to `self.ctx.filter` and then to pass it to be blockstream, so just going with unwrap for now

mangas commented on core/src/subgraph/runner.rs:
```diff
@@ -1041,7 +1102,22 @@ where
-        Ok(Action::Continue)
+        // If any datasource has reached its end block in the range [revert_to_ptr, subgraph_ptr]
+        // then we need to restart the blockstream to reset the TriggerFilter.
+        let needs_restart: bool = self
```
I think this should be a function to improve readability

incrypto32 replied:
Yeah the logic is same but the context in which `has_expired` is used is different.

mangas commented on graph/src/blockchain/mod.rs:
```diff
@@ -255,15 +255,31 @@ pub trait DataSource<C: Blockchain>: 'static + Sized + Send + Sync + Clone {
+    /// **NOTE: DO NOT OVERRIDE THIS METHOD.**
+    /// Use `min_spec_version_inner` for custom logic instead.
+    /// This method is used to determine the minimum spec version required by the data source.
     fn min_spec_version(&self) -> semver::Version {
```
This does not look like the right place for this sort of contraint, maybe the manifest validation would be the right place. This here is leaking the details about spec into data sources

incrypto32 replied:
Got it, i moved this logic to manifest validation

mangas reviewed (approved):
Would be great to see the improvements we discussed privately done before merging but will leave that up to you
